[feat] 일기장 기능 구현

idon1nab opened the pull request:
## 작업 내용
- DiaryService의 CRUD 기능 작성 및 정상 작동 테스트 완료
- DiaryDTO <-> Diary의 원활한 매핑을 위한 modelmapper 설정
- Diary, User 엔티티 수정


## 구현 방법 
- RootConfig 파일에 PropertyMap 설정으로 DiaryDTO와 Diary의 어떤 필드가 매핑 되는지 명시

idon1nab committed:
feat: DiaryDTO 작성

idon1nab committed:
Merge branch 'develop' of https://github.com/Dallili/secretFriends-api into dev/feat/list-diary

idon1nab committed:
feat: DiaryService 작성

idon1nab committed:
feat: DiaryserviceImpl 작성 위한 인터페이스 및 Diary 엔티티 메소드 추가

idon1nab committed:
fix: DiaryServiceImpl modifyPartner() 메소드 수정

idon1nab committed:
feat: DiaryDTO에서 user, partner 수정, DiaryServiceTests testRegister() 작성

idon1nab committed:
feat: testReadOne() 작성

idon1nab committed:
feat: DiaryDTO <-> Diary 매핑을 위한 deep mapping 설정 추가

DiaryDTO의 User user, User partner -> String userID, partnerID로 변경함에 따라 발생한 매핑 오류 문제 해결을 위해 RootConfig에 사용자 정의 매핑 규픽 작성

idon1nab committed:
feat: DiaryService 기본 CRUD 메소드 작성과 테스트 모두 완료

idon1nab committed:
Merge branch 'develop' of https://github.com/Dallili/secretFriends-api into dev/feat/list-diary

idon1nab committed:
Merge branch 'develop' of https://github.com/Dallili/secretFriends-api into dev/feat/list-diary

crHwang0822 commented on src/main/java/org/dallili/secretfriends/service/DiaryServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+        DiaryDTO diaryDTO = modelMapper.map(diary, DiaryDTO.class);
+
+        //좀 더 개발 후에 pageDTO도 함께 다뤄야 함
```
pageDTO도 같이 다루어야 한다는 게 무슨 말이죠 ❓ ❓

crhwang822 commented on src/main/java/org/dallili/secretfriends/service/DiaryServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+        DiaryDTO diaryDTO = modelMapper.map(diary, DiaryDTO.class);
+
+        //좀 더 개발 후에 pageDTO도 함께 다뤄야 함
```
pageDTO도 같이 다루어야 한다는 게 무슨 말이죠 ❓ ❓

idon1nab replied:
사실상 일기장 하나를 read 하는거는 딱히 의미가 없는 것 같아서
진짜 일기를 읽는 기능은 pageDTO 다루는 곳에서 따로 다루거나 해야겠다는 뜻입니당

crhwang822 replied:
일기장 하나를 read 하는 기능은 API명세서에도 없지 않나여 ??
진짜 일기를 읽는 기능은 언니 말대로 page 테이블에서 다루는 거 맞구 API명세서에 있는 일기 조회 기능이 그거라 내가 구현하는 거!

언니가 구현해야하는 일기장 조회는 
1. 유저가 메인 캐비닛 화면에 들어갔을 때 보여지는 다이어리 리스트 (답장O)
2. 유저가 히스토리 캐비닛 화면에 들어갔을 때 보여지는 다이어리 리스트 (활성화)
3. 유저가 가방 화면에 들어갔을 때 보여지는 다이어리 리스트 (비활성화)

이렇게로 알고있는데 아니면 댓 남겨주셔욤 . .

idon1nab replied:
엉 이거 맞아!! 
전에 교재 공부한 거 참고하면서 하다보니 저 메소드가 작성됐는데,
저 당시에 
'이러면 의미가 딱히 없지 않나?? 리턴에서 pageDTO을 같이 다뤄야 하나..?(=사실상 채린이 파트)'
라고 생각해서 주석 남겨놨었다..

근데 지금보니 내가 할 필요가 없는 파트고,, 공부하다 남긴 주석이라 다음에 삭제하고 피알 보낼 예정..

crhwang822 commented on src/main/java/org/dallili/secretfriends/service/DiaryService.java:
```diff
@@ -1,4 +1,20 @@
+
 public interface DiaryService {
+
+    String register(DiaryDTO diaryDTO); // 일기장 등록
```
정말 좋은데 우리 네이밍 컨벤션 정해놨던 것 ,,, 😿

idon1nab replied:
아하 수정할게,,,!! (수정해서 다음  PR에 반영할게욘

crhwang822 commented on src/main/java/org/dallili/secretfriends/service/DiaryServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+        User partner = partnerResult.orElseThrow();
+
+        diary.setPartner(partner);
```
우리 Entity에 Setter 사용 지양해야 한다고 배우지 않았나 ?? 한번 확인 부탁드려욤

idon1nab replied:
엉 지양해야 한다고 알고 있었는데 이 부분은 setter가 꼭 필요한 것 같아서!!
다이어리 처음 만들면 

> 생성 유저 -> user
> 후에 매칭된/코드 입력한 유저 -> partner

로 들어가서 partner가 null 값으로 시작되잖아?? 후에 추가해주려면 setter가 필요한듯..?


(아님 조금 번거롭더라도 build로 만들어서 레포지토리에 save()..? 근데 modelmapper 매핑 설정때문에도 setter가 필요함)

crhwang822 replied:
그 부분을 setter 를 안 쓰고 
Diary 클래스에 메서드를 따로 작성하거나
DiaryRepository에 메서드를 작성하는 식으로 대체하는 게 좋을 것 같우

모델매핑 관련해서는 내가 잘 모르겠다잉

idon1nab replied:
아하 뭔말인지 알겠당 Diary 클래스에 setPartner 대신 decidePartner 따로 메소드 작성할게!

crhwang822 commented on src/main/java/org/dallili/secretfriends/service/DiaryServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+        //log.info("업데이트 전: " + diary);
+
+        diary.updateDiary(diaryDTO.getUpdatedBy(), diaryDTO.getUpdatedAt());
```
diaryDTO.getUpdatedAt() 
이거 업데이트된 시간을 DTO에서 얻고있는데, 
api 요청할 때 업데이트 시간도 프론트단에서 request body에 넣어보내게 할 건가??
난 LocalDateTime.now() 써서 구현해가지구 이것도 통일하면 좋을 듯

idon1nab replied:
아하 지금은 testUpdate()에서 DTO 빌드할 때 LocalDateTime.now() 쓰고 여기에서 시간 따왔는데, (=업데이트 시간을 DTO에서 얻음)
너 말대로 저 코드 파라미터에 바로 LocalDateTime.now() 주는게 더 깔끔할 듯!! 다음 PR에 수정 반영할게여

idon1nab commented:
테스트

idon1nab committed:
feat: findAllDiaries() 작성 및 테스트

idon1nab committed:
feat: 활성화/비활성화 일기장 조회 기능 추가

 DiaryController의 diaryDTOList(Boolean state, String userID)를 통해 활성화 상태와 userID를 조건으로 일기장 목록을 조회

idon1nab committed:
refactor: 컨벤션에 맞게 메소드 명 변경
이전 PR 리뷰 사항 반영

idon1nab committed:
feat: setter 사용 지양을 위한 수정

1. Diary, User에 modelmapper를 위한 setter 분리해서 기록
2. setPartner()-> decidePartner로 변경

idon1nab commented:
## 작업 내용
1. DiaryController에 유저별 활성화 / 비활성화 일기장 목록 조회하는 API 추가 
2. 이전 PR 수정 사항 반영
- LocalDateTime.now() 위치 변경
- setter 관련 수정: setPartner -> decidePartner 변경, changeState 추가 @Setter 삭제
- 컨벤션 규칙 맞추어 service 메소드명 변경
- 주석 삭제

## 구현 방법
diaryDTOList는 활성화 상태 값과 userID 값을 매개변수로 받아서, 해당 유저의 활성화된 or 비활성화된 일기장 목록을 반환한다. 
diaryDTO의 userID와 partnerID를 모두 조사해서 둘 중 한 쪽이라도 포함되어있으면 반환 대상임


## 테스트 내용
- 1. userID가 **user10**이고 **isActivate가 true**인 일기장 리스트
![image](https://github.com/Dallili/secretFriends-api/assets/99960721/c65e83cf-0de7-48a5-b9bd-d21eea2ed564)

- 결과(로그)
![image](https://github.com/Dallili/secretFriends-api/assets/99960721/f8b826b3-4eff-4f46-92f3-453fbd10cff8)
`true상태인 일기장 목록: [DiaryDTO(diaryID=diary10, userID=user10, partnerID=user12, isActivated=true, updatedBy=user10, updatedAt=2024-03-26T15:52:15, color=#000000), DiaryDTO(diaryID=diary101, userID=user1, partnerID=user10, isActivated=true, updatedBy=user10, updatedAt=2024-03-24T16:39:42, color=#123456), DiaryDTO(diaryID=diary102, userID=user1, partnerID=user10, isActivated=true, updatedBy=user10....`

- 2. userID가 **user11**이고 **isActivate가 false**인 일기장 리스트
![image](https://github.com/Dallili/secretFriends-api/assets/99960721/b00afac7-df95-425f-9634-3054265d8207)

- 결과(로그)
![image](https://github.com/Dallili/secretFriends-api/assets/99960721/a5256e25-1ef4-421f-b953-a280cd417658)
`false상태인 일기장 목록: [DiaryDTO(diaryID=diary11, userID=user11, partnerID=user13, isActivated=false, updatedBy=user11, updatedAt=2024-03-26T15:52:15, color=#000000), DiaryDTO(diaryID=diary9, userID=user9, partnerID=user11, isActivated=false, updatedBy=user9, updatedAt=2024-03-26T15:52:15, color=#000000)]`
